File: src/chain-observers/providers/impacted-account-provider.ts
import { TAccountName } from "@hiveio/wax";
import { WorkerBeeIterable } from "../../types/iterator";
import { ImpactedAccountClassifier } from "../classifiers";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { IOperationTransactionPair } from "../classifiers/operation-classifier";
import { TProviderEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export type TImpactedAccountProvided<TAccounts extends Array<TAccountName>> = {
  [K in TAccounts[number]]: WorkerBeeIterable<IOperationTransactionPair>;
};

export interface IImpactedAccountProviderData<TAccounts extends Array<TAccountName>> {
  impactedAccounts: Partial<TImpactedAccountProvided<TAccounts>>;
};

export interface IImpactedAccountProviderOptions {
  accounts: string[];
}

export class ImpactedAccountProvider<
  TAccounts extends Array<TAccountName> = Array<TAccountName>
> extends ProviderBase<IImpactedAccountProviderOptions, IImpactedAccountProviderData<TAccounts>> {
  public readonly accounts = new Set<TAccountName>();

  public pushOptions(options: IImpactedAccountProviderOptions): void {
    for(const account of options.accounts)
      this.accounts.add(account);
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      ImpactedAccountClassifier
    ];
  }

  public get baseStructure(): IImpactedAccountProviderData<TAccounts> {
    return {
      impactedAccounts: {}
    };
  }

  public async provide(data: TProviderEvaluationContext): Promise<IImpactedAccountProviderData<TAccounts>> {
    const result = this.baseStructure;

    const { impactedAccounts } = await data.get(ImpactedAccountClassifier);
    for(const account of this.accounts)
      if (impactedAccounts[account] !== undefined)
        result.impactedAccounts[account] = new WorkerBeeIterable<IOperationTransactionPair>(Array.from(impactedAccounts[account].operations));

    return result as IImpactedAccountProviderData<TAccounts>;
  }
}

File: python/workerbee/chain_observers/providers/impacted_account_provider.py
"""ImpactedAccountProvider — provides operations that impact tracked accounts.

Mirrors src/chain-observers/providers/impacted-account-provider.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .._ordered_set import OrderedSet
from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.impacted_account_classifier import ImpactedAccountClassifier
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import ImpactedAccountsPayload, OperationTransactionPair


class ImpactedAccountProvider(ProviderBase):
    def __init__(self) -> None:
        self.accounts: OrderedSet[str] = OrderedSet()

    def push_options(self, options: Any) -> None:
        for account in options["accounts"]:
            self.accounts.add(account)

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [ImpactedAccountClassifier]

    async def provide(self, data: DataEvaluationContext) -> ImpactedAccountsPayload:
        impacted_data = await data.get(ImpactedAccountClassifier)
        impacted_accounts = impacted_data["impacted_accounts"]

        result: dict[str, list[OperationTransactionPair]] = {}
        for account in self.accounts:
            entry = impacted_accounts.get(account)
            if entry is not None:
                result[account] = list(entry["operations"])

        return {"impacted_accounts": result}
